### PR TITLE
fix(middleware): compute client address

### LINKED ISSUE #12215: `clientAddress` error in Netlify Edge middleware
### Astro Info

```block
Astro                    v4.16.2
Node                     v20.17.0
System                   macOS (arm64)
Package Manager          pnpm
Output                   server
Adapter                  @astrojs/netlify
Integrations             @astrojs/sitemap
                         @astrojs/solid-js
                         astro:db
                         @astrojs/db/file-url
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

When I deploy my site on Netlify with the option `edgeMiddleware: true` I get the following error:

`StaticClientAddressNotAvailable: `Astro.clientAddress` is only available when using `output: 'server'` or `output: 'hybrid'`. Update your Astro config if you need SSR features.`

I have set `output: 'server'` in my config. When I deploy with `edgeMiddleware: false` everything works without errors.

For the reproduction I've created a repository: https://github.com/waspeer/astro-netlify-edge-clientaddress-issue
And deployed it to netlify: https://astro-netlify-edge-clientaddress-issue.netlify.app/
The following error is shown in the edge function logs:

```

StaticClientAddressNotAvailable: `Astro.clientAddress` is only available when using `output: 'server'` or `output: 'hybrid'`. Update your Astro config if you need SSR features.
    at Object.get clientAddress (file:///root/.netlify/v1/edge-functions/middleware/middleware.mjs:2945:13)
    at file:///root/.netlify/v1/edge-functions/middleware/middleware.mjs:1004:23
    at applyHandle (file:///root/.netlify/v1/edge-functions/middleware/middleware.mjs:966:22)
    at file:///root/.netlify/v1/edge-functions/middleware/middleware.mjs:963:12
    at entry_default (file:///root/.netlify/v1/edge-functions/middleware/middleware.mjs:3019:10)
    at file:///root/src/bootstrap/function_chain.ts:426:15
    at AsyncLocalStorage.run (node:async_hooks:231:13)
    at FunctionChain.runFunction (file:///root/src/bootstrap/function_chain.ts:420:28)
    at FunctionChain.run (file:///root/src/bootstrap/function_chain.ts:324:31)
    at handleRequest (file:///root/src/bootstrap/handler.ts:148:34) {
  loc: undefined,
  title: "`Astro.clientAddress` is not available in static mode.",
  hint: "See https://docs.astro.build/en/guides/server-side-rendering/ for more information on how to enable SSR.",
  frame: undefined,
  type: "AstroError",
  name: "StaticClientAddressNotAvailable"
}
```

### What's the expected result?

To be able to access `context.clientAddress` in Netlify edge middleware.

### Link to Minimal Reproducible Example

https://github.com/waspeer/astro-netlify-edge-clientaddress-issue

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

Closes #12215 
Closes PLT-2587

## Testing

Added a unit test.

Existing CI should pass

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/core/middleware/index.ts [hunk 1]
```diff
@@ -85,10 +87,14 @@ function createContext({
 		},
 		url,
 		get clientAddress() {
-			if (clientAddressSymbol in request) {
-				return Reflect.get(request, clientAddressSymbol) as string;
+			if (clientIpAddress) {
+				return clientIpAddress;
 			}
-			throw new AstroError(AstroErrorData.StaticClientAddressNotAvailable);
+			clientIpAddress = getClientIpAddress(request);
```
Comment (matthewp) [inline packages/astro/src/core/middleware/index.ts:93]: I assume this is done here so that we don't need to update each adapter to provide it?

**Review by matthewp — CHANGES_REQUESTED** [review]
Should be able to add a test for this. I think this one would be the place: https://github.com/withastro/astro/blob/main/packages/astro/test/client-address.test.js

File: packages/astro/src/core/middleware/index.ts (hunk 1 above)
Comment (ematipico) [reply packages/astro/src/core/middleware/index.ts:93]: Exactly!

**Comment by ematipico** [comment]
@matthewp the bug is around edge middleware specifically, we don't have those tests in core, I think

**Comment by ematipico** [comment]
@matthewp added a unit test

**Review by matthewp — APPROVED** [review]

**Review by bluwy — APPROVED** [review]

**Pull request merged**